runtime/sema/check_transaction_declaration.go

```diff
@@ -248,36 +303,175 @@ func (checker *Checker) declareTransactionDeclaration(declaration *ast.Transacti
 		transactionType.Parameters = checker.parameters(declaration.ParameterList)
 	}
 
+	members, fields, origins := checker.compositeFieldMembersAndOrigins(
+		transactionType,
+		declaration.DeclarationKind(),
+		declaration.Fields,
+	)
+
+	transactionType.Members = members
+	transactionType.Fields = fields
+	if checker.PositionInfo != nil {
+		checker.PositionInfo.recordMemberOrigins(transactionType, origins)
+	}
+
+	if declaration.Prepare != nil {
+		parameterList := declaration.Prepare.FunctionDeclaration.ParameterList
+		transactionType.PrepareParameters = checker.parameters(parameterList)
+	}
+
+	var fieldDeclarationsByIdentifier map[string]*ast.FieldDeclaration
+	fieldDeclarationByIdentifier := func(identifier string) *ast.FieldDeclaration {
+		if fieldDeclarationsByIdentifier == nil {
+			fieldDeclarationsByIdentifier = make(map[string]*ast.FieldDeclaration, len(declaration.Fields))
+			for _, field := range declaration.Fields {
+				fieldDeclarationsByIdentifier[field.Identifier.Identifier] = field
+			}
+		}
+		return fieldDeclarationsByIdentifier[identifier]
+	}
+
+	roles := &orderedmap.OrderedMap[string, *TransactionRoleType]{}
+	for _, roleDeclaration := range declaration.Roles {
+		transactionRoleType := checker.transactionRoleType(roleDeclaration)
+		checker.Elaboration.SetTransactionRoleDeclarationType(roleDeclaration, transactionRoleType)
+
+		// Ensure roles are not duplicated
+		roleName := roleDeclaration.Identifier.Identifier
+		if _, ok := roles.Get(roleName); ok {
+			checker.report(
+				&DuplicateTransactionRoleError{
+					Name: roleName,
+					Range: ast.NewRangeFromPositioned(
+						checker.memoryGauge,
+						roleDeclaration.Identifier,
+					),
+				},
+			)
+		} else {
+			roles.Set(roleName, transactionRoleType)
+		}
+
+		// Ensure roles and fields do not clash
+		if _, ok := members.Get(roleName); ok {
+			field := fieldDeclarationByIdentifier(roleName)
```

Comment:
Seems the `field` is only used to get the field's Identifier. The same info should be available via `member.Identifier`

Reply:
With that, I think it is not needed to maintain a separate `roles` mapping, given that roles also are added to members, "duplicate roles" also can be checked here along with fields. The reported error can be switched based on the `member.DeclarationKind`

Reply:
Good idea! Refactored in 170f96aaa17becb8b72d5f3c0e63748d0b513028

Reply:
And further refactored in 20d50449f2a77150143b620b2cc9e1db60814601